File: frontend/app/actions/departments.ts
"use server";

import { revalidatePath } from "next/cache";
import { apiFetch } from "@/lib/api";
import type { DepartmentItem, DeptStatsReport } from "@/lib/types";
import { responseErrorMessage } from "./errors";

export async function listDepartments(workspaceSlug: string): Promise<DepartmentItem[]> {
  const res = await apiFetch(`/admin/v1/departments?workspace=${encodeURIComponent(workspaceSlug)}`);
  if (!res.ok) throw new Error(`Failed to load departments (${res.status})`);
  return res.json() as Promise<DepartmentItem[]>;
}

export async function listDeptStats(workspaceSlug: string): Promise<DeptStatsReport> {
  const res = await apiFetch(`/admin/v1/stats/workspaces/${encodeURIComponent(workspaceSlug)}/departments`);
  if (!res.ok) throw new Error(`Failed to load department stats (${res.status})`);
  return res.json() as Promise<DeptStatsReport>;
}

export async function createDepartment(
  workspaceSlug: string,
  name: string,
): Promise<{ ok: true; dept: DepartmentItem } | { ok: false; error: string }> {
  let res: Response;
  try {
    res = await apiFetch(`/admin/v1/workspaces/${encodeURIComponent(workspaceSlug)}/departments`, {
      method: "POST",
      body: JSON.stringify({ name }),
    });
  } catch (e) {
    return { ok: false, error: (e as Error).message };
  }

  if (!res.ok) {
    return { ok: false, error: await responseErrorMessage(res, `Create failed (${res.status})`) };
  }

  const dept = (await res.json()) as DepartmentItem;
  revalidatePath("/dashboard/departments");
  return { ok: true, dept };
}

export async function renameDepartment(
  workspaceSlug: string,
  deptSlug: string,
  name: string,
): Promise<{ ok: true; dept: DepartmentItem } | { ok: false; error: string }> {
  let res: Response;
  try {
    res = await apiFetch(
      `/admin/v1/workspaces/${encodeURIComponent(workspaceSlug)}/departments/${encodeURIComponent(deptSlug)}`,
      { method: "PATCH", body: JSON.stringify({ name }) },
    );
  } catch (e) {
    return { ok: false, error: (e as Error).message };
  }

  if (!res.ok) {
    return { ok: false, error: await responseErrorMessage(res, `Rename failed (${res.status})`) };
  }

  const dept = (await res.json()) as DepartmentItem;
  revalidatePath("/dashboard/departments");
  return { ok: true, dept };
}

export async function deleteDepartment(
  workspaceSlug: string,
  deptSlug: string,
): Promise<{ ok: true } | { ok: false; error: string }> {
  let res: Response;
  try {
    res = await apiFetch(
      `/admin/v1/workspaces/${encodeURIComponent(workspaceSlug)}/departments/${encodeURIComponent(deptSlug)}`,
      { method: "DELETE" },
    );
  } catch (e) {
    return { ok: false, error: (e as Error).message };
  }

  if (!res.ok) {
    return { ok: false, error: await responseErrorMessage(res, `Delete failed (${res.status})`) };
  }

  revalidatePath("/dashboard/departments");
  return { ok: true };
}

File: server/app/routers/admin/departments.py
from fastapi import APIRouter, HTTPException, Query, status

from app.schemas.admin.departments import (
    DepartmentCreate,
    DepartmentDeleteResponse,
    DepartmentItem,
    DepartmentUpdate,
)
from app.services import admin_service

router = APIRouter()


@router.get("/departments", response_model=list[DepartmentItem])
def list_departments(workspace: str | None = Query(default=None)):
    try:
        return admin_service.list_departments(workspace_slug=workspace)
    except admin_service.WorkspaceNotFound:
        return []


@router.post(
    "/workspaces/{workspace_slug}/departments",
    response_model=DepartmentItem,
    status_code=status.HTTP_201_CREATED,
)
def create_department(workspace_slug: str, body: DepartmentCreate):
    try:
        return admin_service.create_department(workspace_slug, body.name)
    except admin_service.WorkspaceNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except admin_service.DuplicateSlug as exc:
        raise HTTPException(status_code=409, detail="Department slug already exists") from exc


@router.patch(
    "/workspaces/{workspace_slug}/departments/{dept_slug}",
    response_model=DepartmentItem,
)
def rename_department(workspace_slug: str, dept_slug: str, body: DepartmentUpdate):
    try:
        return admin_service.rename_department(workspace_slug, dept_slug, body.name)
    except admin_service.WorkspaceNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except admin_service.DeptNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.delete(
    "/workspaces/{workspace_slug}/departments/{dept_slug}",
    response_model=DepartmentDeleteResponse,
)
def delete_department(workspace_slug: str, dept_slug: str):
    try:
        return admin_service.delete_department(workspace_slug, dept_slug)
    except admin_service.WorkspaceNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except admin_service.DeptNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
